### PR TITLE
Checkpoint end to end test

### PR DESCRIPTION
**Pull request by asonnino**
An end-to-end test for the checkpointing logic. 

* The active validator runs the checkpoint engine in the background
* We submit a few transactions to the system through the authority aggregator
* The transactions are executed and eventually become part of a checkpoint fragment
* The fragments are sequenced by each authority
* Eventually the authorities increase their next expected checkpoint sequence number

### CODE REVIEW

File: crates/sui-core/src/authority_active.rs [hunk 1]
```diff
@@ -190,11 +190,17 @@ where
...
+            .await
     }
 
     /// Spawn all active tasks.
-    pub async fn spawn_active_processes(self, gossip: bool, checkpoint: bool) {
+    pub async fn spawn_active_processes(
+        self,
+        gossip: bool,
+        checkpoint: bool,
+        checkpoint_process_control: CheckpointProcessControl,
```
Comment (asonnino) [inline crates/sui-core/src/authority_active.rs:202]: This allows us to input a custom `CheckpointProcessControl` with smaller timeouts than the defaults (useful for integration tests).

File: crates/sui/tests/checkpoints_tests.rs [hunk 2]
```diff
@@ -80,3 +86,69 @@ async fn sequence_fragments() {
         sleep(Duration::from_millis(10)).await;
     }
 }
+
+#[tokio::test]
+async fn end_to_end() {
```
Comment (asonnino) [inline crates/sui/tests/checkpoints_tests.rs:91]: That is the new test. Is it the end to end test you were expecting to see (or did I miss something)?

Comment (asonnino) [reply crates/sui/tests/checkpoints_tests.rs:91]: @gdanezis

File: crates/sui/tests/shared_objects_tests.rs [hunk 3]
```diff
@@ -495,7 +495,7 @@ async fn shared_object_on_gateway() {
     // the handles (or the authorities will stop).
     let configs = test_authority_configs();
     let _handles = spawn_test_authorities(gas_objects.clone(), &configs).await;
-    let clients = create_authority_aggregator(configs.validator_set());
+    let clients = test_authority_aggregator(&configs);
```
Comment (asonnino) [inline crates/sui/tests/shared_objects_tests.rs:498]: Just a rename for consistency (trying to get all testing facilities in the same place)

File: crates/sui/tests/checkpoints_tests.rs (hunk 2 above)
Comment (asonnino) [reply crates/sui/tests/checkpoints_tests.rs:91]: good points

File: crates/sui/tests/checkpoints_tests.rs [hunk 4]
```diff
@@ -80,3 +93,244 @@ async fn sequence_fragments() {
...
+                        state
+                            .lock()
+                            .checkpoint_contents
+                            .get(&(checkpoint_sequence, i))
+                            .unwrap()
+                    })
+                    .map(|x| x.transaction)
+                    .collect::<HashSet<_>>()
+            })
+            .collect::<HashSet<_>>();
```
[the remark this answers is not in the document]
Comment (asonnino) [reply crates/sui/tests/checkpoints_tests.rs]: Cool, then I isolated this check on a separate function so we will just have to update it in one place

File: crates/test-utils/src/authority.rs [hunk 5]
```diff
@@ -70,3 +84,80 @@ pub fn create_authority_aggregator(
...
+            .await
+            .unwrap();
+        responses.push(reply);
+    }
+    responses
+}
+
+/// Keep submitting the certificates of a shared-object transaction until it is sequenced by
+/// at least one consensus node. We use the loop since some consensus protocols (like Tusk)
+/// may drop transactions. The certificate is submitted to every Sui authority.
```
Comment (gdanezis) [inline crates/test-utils/src/authority.rs:115]: But is the resubmission not usually done by the validator?

**Review by gdanezis — APPROVED** [review]
Nice one - I think we have what we need as a first e2e test. What are we missing to turn it on?

File: crates/test-utils/src/authority.rs (hunk 5 above)
Comment (asonnino) [reply crates/test-utils/src/authority.rs:115]: Yes it is. These facilities are just test clients

File: crates/sui/tests/checkpoints_tests.rs [hunk 6]
```diff
@@ -80,3 +123,211 @@ async fn sequence_fragments() {
         sleep(Duration::from_millis(10)).await;
     }
 }
+
+#[tokio::test]
```
[the remark this answers is not in the document]
Comment (asonnino) [reply crates/sui/tests/checkpoints_tests.rs:127]: no it doesn't work (don't know why yet, but it seems that most of our tokio tests don't work with simulated time)

File: crates/sui-core/src/authority_active.rs (hunk 1 above)
Comment (asonnino) [reply crates/sui-core/src/authority_active.rs:216]: fixing in a separate PR (this one auto-merged)

**Pull request merged**